src/coding-guidelines/expressions.rst

```diff
@@ -83,6 +83,61 @@ Expressions
          fn with_base(_: &Base) { ... }
 
 
+.. guideline:: Do not divide by 0
+   :id: gui_kMbiWbn8Z6g5
+   :category: mandatory
+   :status: draft
+   :release: latest
+   :fls: fls_Q9dhNiICGIfr
+   :decidability: undecidable
+   :scope: system
+   :tags: numerics
+
+   This guideline applies when unsigned integer or two’s complement division is performed during the
+   evaluation of an `ArithmeticExpression
+   <https://rust-lang.github.io/fls/expressions.html#arithmetic-expressions>`_.
+
+   This includes the evaluation of a `RemainderExpression
+   <https://rust-lang.github.io/fls/expressions.html#syntax_remainderexpression>`_, which uses unsigned integer or two's
+   complement division.
+
+   This rule does not apply to evaluation of a :std:`core::ops::Div` trait on types other than `integer
+   types <https://rust-lang.github.io/fls/types-and-traits.html#integer-types>`_.
+
+   .. rationale::
+      :id: rat_h84NjY2tLSBW
+      :status: draft
+
+      Integer division by zero results in a panic, which is an abnormal program state and may terminate the
+      process. The use of :std:`std::num::NonZero` as the divisor is a recommended way to avoid the
+      undecidability of this guideline.
+
+   .. non_compliant_example::
+      :id: non_compl_ex_LLs3vY8aGz0F
+      :status: draft
+
+      When the division is performed, the right operand is evaluated to zero and the program panics.
+
+      .. code-block:: rust
+
+         let x = 0;
+         let x = 5 / x;
+
+   .. compliant_example::
+      :id: compl_ex_Ri9pP5Ch3kbb
+      :status: draft
+
+      There is no compliant way to perform integer division by zero. A checked division will prevent any
+      division by zero from happening. The programmer can then handle the returned :std:``std::option::Option``.
+
+      .. code-block:: rust
+
+         let result = match 5u8.checked_div(0) {
+            None => 0
+            Some(r) => r
+         };
+
```

Comment:
I would like to add another compliant solution here that looks like this:

Another compliant solution is to avoid division by zero:

.. code-block:: rust
 
if (x != 0) {
         let x = 5 / x;
}

Reply:
This is a very reasonable way to comply as well.

If the blocks are small enough, I'm pretty sure both LLVM and GCC are able to optimize away the zero-division-abort that would otherwise be present (by way of value range analysis on the divisor).

Reply:
I agree with your analysis @felix91gr, however I will point out that this Guideline is marked as "Undecidable" already, and follows the CERT style. As soon as you have that designation it is up to the quality of the static analysis tool's implementation as to whether it can or should do anything. The counterpart guideline in #136 is the decidable one.

I also do agree that,

```rust
if (x != 0) {
   let x = 5 / x;
}
else {
   let x = 0;
}
```

is compliant. However I'm not sure I want to promote that in our examples section, as that makes clippy developers sweat

Reply:
You're totally right. This guideline is CERT-style, so it's about defects rather than subsets. 100% right.

Reply:
After some more thought, I added the original example in 10efe3a. I did caveat it with a cautionary note in the description.